### ISSUE TITLE
Content Security Policy

### ISSUE THREAD

**Issue by denchen** [maladr0it/react-truncate-list#5]
Once again, great library. I'm using it a lot, but I've run into an issue and there is an easy fix / enhancement, and I can submit a PR, but I want to run it by you first.

### Background

My React app uses a Node/Express backend, and I've been trying to tighten up my [Content Security Policy](https://content-security-policy.com/) settings. Namely, for the `style-src` rule, previously I had it set with ["unsafe-inline"](https://content-security-policy.com/unsafe-inline/) so the CSS-in-JS library I use could dynamically insert styles into the DOM. I removed that rule and started using [nonce](https://content-security-policy.com/nonce/) for CSS-in-JS, and it worked great. But now the issue is that this library won't work without "unsafe-inline"

### Problem

The error in question:

```
Refused to apply inline style because it violates the following Content Security Policy directive: "style-src 'self' 'nonce-abcde12335ABCDE09876'". Either the 'unsafe-inline' keyword, a hash ('sha256-47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU='), or a nonce ('nonce-...') is required to enable inline execution.
```

This is because of how CSS is applied in your code:

```jsx
import "./styles.css";
```

But the way the library is built, with `rollup-plugin-postcss`, the production build produces the following:

```jsx
function styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = ".TruncatedList {\n  box-sizing: border-box;\n  padding-left: 0;\n  margin: 0;\n  list-style: none;\n  overflow: auto;\n}\n\n.TruncatedList * {\n  box-sizing: inherit;\n}\n";
styleInject(css);
```

It's the `appendChild()` that is blocked by my Content Security Policy.

### Solution

I have a solution that involves passing in an optional `nonce` prop so this library is allowed to insert styles. However, my solution had to change the way the base CSS is inserted, by using a `<style>` tag. If a `nonce` exists, then it'll be used within the tag with `<style nonce={nonce}>`. The biggest difference is that the CSS that was previously within `styles.css` is now inlined within `index.js`. However, since the CSS is pretty short and lightweight, I found this to be an acceptable change.

If it's okay with you, I can submit a PR for your review.

**Comment by maladr0it** [comment]
This sounds suitable to me; considering the very small amount of css, using
a <style> tag with appropriately named classes would also have the
advantage of reducing the bundle size.

Go ahead and submit a PR and I'll look it over ASAP.

**Comment by maladr0it** [comment]
Sorry for the delay. Upon reflection, how would you feel about the library instead just providing the .css file and you import it yourself? I am just wary of making the API more complicated than it needs to be, and importing the required .css is a common practice for consumers of a library.

**Comment by denchen** [comment]
Sorry for my own delayed response.

So what you proposed was actually my initial approach. I basically set `extract: true` in the `rollup-plugin-postcss` options, which builds this library with the CSS remaining in its own file. However, I felt this would be an inconvenience -- albeit a small one -- to have your users have to include this CSS, especially when they didn't have to before.

That being said, doing it this way would probably allow for more styling control. Instead of including your CSS, which is pretty basic, the user could just insert their own under the `.TruncatedList` CSS namespace. And if you choose to extract your CSS, then this is probably what I will do; I'll just copy your CSS into my global CSS that I set up with CSS-in-JS.

Tangent: I kinda feel regardless of what you decide here, I think having `.react-truncate-list` as the base class instead of `.TruncatedList` would be better since it would match the library name. Thoughts?

My proposal is an added API, yes, but it's entirely optional. If a user knows nothing of CSP and doesn't use it, then that user can safely ignore it. I feel in terms of complexity, it's almost an equal tradeoff between needing to include an extra CSS file vs. using a CSP-specific option.

One thing I did consider in the implementation was to make it all automatic. The library would search for `<meta name="csp-nonce">` and use that value (this is what Material UI does, if I'm not mistaken), so the user wouldn't have to deal with it. Or it would search for a global environment value (I _think_ Semantic UI uses this approach). But I quickly realized this would bloat the code way too much, and it fell outside the scope of this library.

One thing I don't know is either solution is appropriate for SSR. I don't use SSR in my app, and I have no experience with it.

In summary, with regards to my app and how I'm using your library, obviously this PR works for my scenario, but having the CSS as a separate file would work equally well for my use-case.

**Comment by maladr0it** [comment]
Please check out my recent changes in v0.3.4, the styles are imported from:
`react-truncate-list/dist/styles.css`

I think this gives the user the most flexibility in how they add the base styles, and it also reduces the amount of JS by roughly half by not including the CSS injection code.

If you are okay with this solution, I will close your PR.

**Comment by denchen** [comment]
Hey,

So I'm okay with your PR, and I have noticed that you changed the base class to `.react-truncate-list` 👍 As I mentioned,I do have the lexibility of either importing your CSS or creating my own CSS.Either way, the main issue of CSP violation would be resolved. I'll let you know if I have any further issues.
